sdk/go/common/resource/config/object.go

```diff
@@ -0,0 +1,487 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+var errSecureReprReserved = errors.New(`maps with the single key "secure" are reserved`)
+
+// object is the internal object representation of a single config value. All operations on Value first decode the
+// Value's string representation into its object representation. Secure strings are stored in objects as ciphertext.
+type object struct {
+	value  any
+	secure bool
+}
+
+// objectType describes the types of values that may be stored in the value field of an object
+type objectType interface {
+	bool | int64 | float64 | string | []object | map[string]object
+}
+
+// newObject creates a new object with the given representation.
+func newObject[T objectType](v T) object {
+	return object{value: v}
+}
+
+// newSecureObject creates a new secure object with the given ciphertext.
+func newSecureObject(ciphertext string) object {
+	return object{value: ciphertext, secure: true}
+}
+
+// Secure returns true if the receiver is a secure string or a composite value that contains a secure string.
+func (c object) Secure() bool {
+	switch v := c.value.(type) {
+	case []object:
+		for _, v := range v {
+			if v.Secure() {
+				return true
+			}
+		}
+		return false
+	case map[string]object:
+		for _, v := range v {
+			if v.Secure() {
+				return true
+			}
+		}
+		return false
+	case string:
+		return c.secure
+	default:
+		return false
+	}
+}
+
+// Decrypt decrypts any ciphertexts within the object and returns appropriately-shaped Plaintext values.
+func (c object) Decrypt(ctx context.Context, decrypter Decrypter) (Plaintext, error) {
+	switch v := c.value.(type) {
+	case bool:
+		return NewPlaintext(v), nil
+	case int64:
+		return NewPlaintext(v), nil
+	case float64:
+		return NewPlaintext(v), nil
+	case string:
+		if !c.secure {
+			return NewPlaintext(v), nil
+		}
+		plaintext, err := decrypter.DecryptValue(ctx, v)
+		if err != nil {
+			return Plaintext{}, err
+		}
+		return NewSecurePlaintext(plaintext), nil
+	case []object:
+		vs := make([]Plaintext, len(v))
+		for i, v := range v {
+			pv, err := v.Decrypt(ctx, decrypter)
+			if err != nil {
+				return Plaintext{}, err
+			}
+			vs[i] = pv
+		}
+		return NewPlaintext(vs), nil
+	case map[string]object:
+		vs := make(map[string]Plaintext, len(v))
+		for k, v := range v {
+			pv, err := v.Decrypt(ctx, decrypter)
+			if err != nil {
+				return Plaintext{}, err
+			}
+			vs[k] = pv
+		}
+		return NewPlaintext(vs), nil
+	default:
+		contract.Failf("unexpected value of type %T", v)
+		return Plaintext{}, nil
+	}
+}
+
+// Get gets the member value at path. The path to the receiver is prefix.
+func (c object) Get(prefix, path resource.PropertyPath) (object, bool, error) {
```

Comment:
The boolean appears to indicate success. Consider using a named return there for documentation.

```
(_ object, ok bool, err error)
```